src/quokka_editor_back/tests/utils/test_ot.py

```diff
@@ -0,0 +1,340 @@
+    # Given
+    prev_op = OperationSchema(
+        from_pos=PosSchema(ch=0, line=0),
+        to_pos=PosSchema(ch=0, line=0),
+        text=["text"],
+        type=OperationType.DELETE,
+        revision=0,
+    )
+    new_op = OperationSchema(
+        from_pos=new_op_from_pos,
+        to_pos=PosSchema(ch=0, line=0),
+        text=["text"],
+        type=new_op_type,
+        revision=1,
+    )
+
+    # When
+    result = transform(new_op, prev_op)
+
+    # Then
+    assert result == OperationSchema(
+        from_pos=PosSchema(line=new_op.from_pos.line - 1, ch=new_op.from_pos.ch),
+        to_pos=new_op.to_pos,
+        text=new_op.text,
+        type=new_op.type,
+        revision=new_op.revision,
+    )
+
+
+@pytest.mark.parametrize(
+    "new_op_type, prev_op_from_pos",
+    [
+        (
+            OperationType.INPUT,
+            PosSchema(ch=0, line=1),
+        ),
+        (
+            OperationType.INPUT,
+            PosSchema(ch=0, line=0),
+        ),
+        (
+            OperationType.INPUT,
+            PosSchema(ch=1, line=0),
+        ),
+        (
+            OperationType.UNDO,
+            PosSchema(ch=0, line=1),
+        ),
+        (
+            OperationType.UNDO,
+            PosSchema(ch=0, line=0),
+        ),
+        (
+            OperationType.UNDO,
+            PosSchema(ch=1, line=0),
+        ),
+        (
+            OperationType.PASTE,
+            PosSchema(ch=0, line=1),
+        ),
+        (
+            OperationType.PASTE,
+            PosSchema(ch=0, line=0),
+        ),
+        (
+            OperationType.PASTE,
+            PosSchema(ch=1, line=0),
+        ),
+    ],
+)
+def test_transform_prev_op_delete_with_specific_conditions(
+    new_op_type, prev_op_from_pos
+):
+    # Given
+    prev_op = OperationSchema(
+        from_pos=prev_op_from_pos,
+        to_pos=PosSchema(ch=0, line=0),
+        text=["text"],
+        type=OperationType.DELETE,
+        revision=0,
+    )
+    new_op = OperationSchema(
+        from_pos=PosSchema(ch=0, line=0),
+        to_pos=PosSchema(ch=0, line=0),
+        text=["text"],
+        type=new_op_type,
+        revision=1,
+    )
+
+    # When
+    result = transform(new_op, prev_op)
+
+    # Then
+    assert result == new_op
+
+
+@pytest.mark.parametrize(
+    "prev_op_type",
+    [
+        OperationType.INPUT,
+        OperationType.UNDO,
+        OperationType.PASTE,
+    ],
+)
+def test_new_op_delete(prev_op_type):
+    # Given
+    prev_op = OperationSchema(
+        from_pos=PosSchema(ch=0, line=0),
+        to_pos=PosSchema(ch=4, line=0),
+        text=["text"],
+        type=prev_op_type,
+        revision=0,
+    )
+    new_op = OperationSchema(
+        from_pos=PosSchema(ch=4, line=0),
+        to_pos=PosSchema(ch=8, line=0),
+        text=["text"],
+        type=OperationType.DELETE,
+        revision=1,
+    )
+
+    # When
+    result = transform(new_op, prev_op)
+
+    # Then
+    # Need explanation
```

Comment:
Write a question here what needs to be explained

Reply:
It is outdated now. You explained it to me on Sunday